kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/ksp/TargetTypes.kt

```diff
@@ -0,0 +1,283 @@
+      name = name,
+      index = index,
+      type = parameter.type.toTypeName(typeParameterResolver),
+      hasDefault = parameter.hasDefault,
+      qualifiers = parameter.qualifiers(resolver),
+      jsonName = parameter.jsonName(resolver)
+    )
+  }
+
+  val kmConstructorSignature: String = resolver.mapToJvmSignature(primaryConstructor)
+    ?: run {
+      logger.error("No primary constructor found.", primaryConstructor)
+      return null
+    }
+  return TargetConstructor(
+    parameters, primaryConstructor.getVisibility().asKModifier(),
+    kmConstructorSignature
+  )
+}
+
+private fun KSAnnotated?.qualifiers(resolver: Resolver): Set<AnnotationSpec> {
+  if (this == null) return setOf()
+  val jsonQualifierType = resolver.getClassDeclarationByName<JsonQualifier>().asType()
+  return annotations
+    .filter {
+      it.annotationType.resolve().declaration.findAnnotationWithType(jsonQualifierType) != null
+    }
+    .mapTo(mutableSetOf()) {
+      it.toAnnotationSpec(resolver)
+    }
+}
+
+private fun KSAnnotated?.jsonName(resolver: Resolver): String? {
+  if (this == null) return null
+  val jsonType = resolver.getClassDeclarationByName<Json>().asType()
+  return findAnnotationWithType(jsonType)?.getMember<String>("name")
+}
+
+private fun declaredProperties(
+  constructor: TargetConstructor,
+  originalType: KSClassDeclaration,
+  classDecl: KSClassDeclaration,
+  resolver: Resolver,
+  typeParameterResolver: TypeParameterResolver,
+): Map<String, TargetProperty> {
+
+  val result = mutableMapOf<String, TargetProperty>()
+  for (property in classDecl.getDeclaredProperties()) {
+    val initialType = property.type.resolve()
+    val resolvedType = if (initialType.declaration is KSTypeParameter) {
+      property.asMemberOf(originalType.asType())
+    } else {
+      initialType
+    }
+    val propertySpec = property.toPropertySpec(resolver, resolvedType, typeParameterResolver)
+    val name = propertySpec.name
+    val parameter = constructor.parameters[name]
+    result[name] = TargetProperty(
+      propertySpec = propertySpec,
+      parameter = parameter,
+      visibility = property.modifiers.map { KModifier.valueOf(it.name) }.visibility(),
+      jsonName = parameter?.jsonName ?: property.jsonName(resolver)
+        ?: name.escapeDollarSigns()
+    )
+  }
+
+  return result
+}
+
+private fun KSPropertyDeclaration.toPropertySpec(
+  resolver: Resolver,
+  resolvedType: KSType,
+  typeParameterResolver: TypeParameterResolver,
+): PropertySpec {
+  return PropertySpec.builder(
+    name = simpleName.getShortName(),
+    type = resolvedType.toTypeName(typeParameterResolver).unwrapTypeAlias()
+  )
+    .mutable(isMutable)
+    .addModifiers(modifiers.map { KModifier.valueOf(it.name) })
+    .apply {
+      if (hasAnnotation(resolver.getClassDeclarationByName<Transient>().asType())) {
+        addAnnotation(Transient::class)
+      }
+      addAnnotations(
+        this@toPropertySpec.annotations
```

Comment:
Don't know if there's an intention to support this, but I noticed a different from kapt if you qualify the annotation with `get`/`set`, ex:

```kotlin
class ClassWithQualiifer(@get:MyQualifier val a: Int)
```
will see the qualifier in kapt, but ignore it in ksp. 

If you do want to look at these, you need to check `getter.annotations` and/or `setter.annotations` as well.